### PR TITLE
Timeseries cleanup, groupBy support and processing chains

### PR DESCRIPTION
**Pull request by pjm17971**
**General**

Enables chrome dev tools custom formatting so that printing out a pond object doesn't drop you into the ugly immutable structures that back Pond.

**Timeseries**

Simplify the `TimeSeries` code by removing the `Series` base class. Originally the base class was going to support time, timerange and index based series as three sub-classes. But it's much simpler to just keep the three different types of lists as members on the `TimeSeries` itself.

The TimeSeries may either have a list of `_times`, `_timeRanges` or `_indexes`. The rest of the code can now use the existence of those lists to perform operations that are dependent on the type of the series, rather the guess by type: string, list, or date. This means this (very easy) decision can often be made once and not per row, so this should be faster.

TODO: Still needs testing around TimeRange based TimeSeries, this is really fully supported yet.

Event changes, mostly to support processing changes and groupBy funct… …
…ionality.

**Processing**
 - Major change: Added processing chains, and generally cleaned up a lot of the processing operators. Here's an example:
```
    const processor = Processor()
        .groupBy("channelName")
        .aggregate("5m", avg)
        .collect("1d", false, timeseries => {
            // --> Output timeseries representing 1 day of 5 min averages
        });
```
 - uniform api for using processing function
 - ability to emit always or next (bucket) out of aggregators and collectors.

**Events**
 - now uses Immutable.map from the top level down
 - added combine(), which takes a list of events, which columns to work with, and a reducer function to apply.
 - sum(), avg() are now implemented with combine().
 - events have keys now that you can set and get to support groupBy functionality in the processor chains
 - added a more uniform approach to reducing a collection of events using map(), reduce() and mapReduce() functions.

### CODE REVIEW

File: docs/events.md [hunk 1]
```diff
@@ -120,6 +137,22 @@ event.get("b") // {in: 654, out: 223};
 
 ### Mutation API
 
+#### getKey(key)
```
Comment (jdugan1024) [inline docs/events.md]: This should be setKey, no?

File: src/aggregator.js [hunk 2]
```diff
@@ -8,72 +8,103 @@
...
+import _ from "underscore";
 import Generator from "./generator";
 
+/**
+ * An aggregator takes the following options:
+ *
+ *     - 'window'    -     size of the window to aggregate over (e.g. "5m")
+ *     - 'operator'  -     function (e.g. avg)
+  *    - 'fieldSpec' -     (optional) describes what part of the events to
+ *                         include in the aggregation. Maybe a string, list
```
Comment (jdugan1024) [inline src/aggregator.js]: May be, rather than maybe.

File: src/aggregator.js [hunk 3]
```diff
@@ -8,72 +8,103 @@
...
-        if (thisBucketIndex !== currentBucketIndex) {
-            if (this._bucket) {
-                this._bucket.aggregate(this._processor, event => {
-                    if (this._observer) {
-                        this._observer(this._bucket.index(), event);
-                    }
-                });
-            }
-            this._bucket = this._generator.bucket(d);
+    constructor(options, observer) {
```
Comment (jdugan1024) [inline src/aggregator.js]: I don't think docs got updated for new constructor signature, looks like options is a object now, but docs have discrete arguments.

File: src/processor.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+    derivative(window, fieldSpec) {
+        if (this._end) {
+            throw new Error("Cannot chain a derivative calculator after the chain has ended.");
+        }
+        const derivative = new Derivative({
+            window,
+            fieldSpec
+        });
+        this._processingList.push(derivative);
```
Comment (jdugan1024) [inline src/processor.js]: this pattern happens all of the time here, push onto processing list and wire up the onEmit for _current and then update. I wonder if it's worth factoring that out into a method.

**Comment by jdugan1024** [comment]
+1 Looks like good stuff all the way around.  My only concern is that it appears the docs may not have been updated fully.  I flagged it in my comments, but I didn't check all the other docs.

**Comment by pjm17971** [comment]
Yep, more time updating the docs is on my todo list. I plan to add/improve them before I version up. Thanks for taking a look.

**Pull request merged**